File: tests/test_gcal_sync.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"time"

	"golang.org/x/oauth2/google"
	"google.golang.org/api/calendar/v3"
	"google.golang.org/api/option"
)

func main() {
	// Booking details
	bookingID := "c0ae2e14-8a1a-4ecf-84ef-3557d7d1b345"
	startTime := "2026-03-30T10:00:00-03:00"
	endTime := "2026-03-30T11:00:00-03:00"
	title := "Cita Médica - Test Booking"
	description := "Test booking created via CLI"
	calendarID := "primary"

	fmt.Println("═══════════════════════════════════════════════════════════")
	fmt.Println("  GCAL SYNC - TEST BOOKING")
	fmt.Println("═══════════════════════════════════════════════════════════")
	fmt.Printf("Booking ID: %s\n", bookingID)
	fmt.Printf("Start: %s\n", startTime)
	fmt.Printf("End: %s\n", endTime)
	fmt.Printf("Title: %s\n", title)
	fmt.Println("═══════════════════════════════════════════════════════════")
	fmt.Println()

	// Load credentials
	credsPath := os.Getenv("GCAL_CREDENTIALS_PATH")
	if credsPath == "" {
		homeDir, _ := os.UserHomeDir()
		credsPath = filepath.Join(homeDir, ".secrets_wm", "booking-sa-key.json")
	}

	fmt.Printf("Loading credentials from: %s\n", credsPath)
	credsJSON, err := ioutil.ReadFile(credsPath)
	if err != nil {
		fmt.Printf("❌ Failed to read credentials: %v\n", err)
		os.Exit(1)
	}
	fmt.Println("✅ Credentials loaded")

	// Create context
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	// Create credentials
	creds, err := google.CredentialsFromJSON(ctx, credsJSON, calendar.CalendarScope)
	if err != nil {
		fmt.Printf("❌ Failed to parse credentials: %v\n", err)
		os.Exit(1)
	}
	fmt.Println("✅ Credentials parsed")

	// Create Calendar service
	srv, err := calendar.NewService(ctx, option.WithCredentials(creds))
	if err != nil {
		fmt.Printf("❌ Failed to create Calendar service: %v\n", err)
		os.Exit(1)
	}
	fmt.Println("✅ Calendar service created")

	// Create event
	event := &calendar.Event{
		Summary:     title,
		Description: description,
		Start: &calendar.EventDateTime{
			DateTime: startTime,
			TimeZone: "America/Argentina/Buenos_Aires",
		},
		End: &calendar.EventDateTime{
			DateTime: endTime,
			TimeZone: "America/Argentina/Buenos_Aires",
		},
	}

	fmt.Println()
	fmt.Println("Creating GCal event...")
	createdEvent, err := srv.Events.Insert(calendarID, event).Do()
	if err != nil {
		fmt.Printf("❌ Failed to create event: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("✅ Event created successfully!")
	fmt.Println()
	fmt.Println("EVENT DETAILS:")
	fmt.Printf("  Event ID: %s\n", createdEvent.Id)
	fmt.Printf("  Summary: %s\n", createdEvent.Summary)
	fmt.Printf("  Start: %s\n", createdEvent.Start.DateTime)
	fmt.Printf("  End: %s\n", createdEvent.End.DateTime)
	fmt.Printf("  HTML Link: %s\n", createdEvent.HtmlLink)
	fmt.Println()

	// Save event ID for DB update
	eventJSON, _ := json.Marshal(map[string]string{
		"event_id":   createdEvent.Id,
		"html_link":  createdEvent.HtmlLink,
		"status":     "success",
	})
	fmt.Printf("Event JSON: %s\n", string(eventJSON))
	fmt.Println()
	fmt.Println("✅ GCAL SYNC COMPLETE!")
}

File: tests/test-gcal-sync.ts
#!/usr/bin/env tsx
/*
 * Test GCal sync for the booking we just created
 */

const BOOKING_ID = 'bee36574-fbd0-42e0-aa83-c34677d0f0b2';
const DB_URL = process.env['DATABASE_URL'] ?? '';
const GCAL_ACCESS_TOKEN = process.env['GCAL_ACCESS_TOKEN'] ?? '';
const GCAL_PROVIDER_CALENDAR_ID = process.env['GCAL_PROVIDER_CALENDAR_ID'] ?? '';

if (!DB_URL) {
  console.error('ERROR: DATABASE_URL not set');
  process.exit(1);
}

async function main() {
  console.log('📅 Testing GCal sync for booking:', BOOKING_ID);
  console.log('━'.repeat(50));
  
  // Step 1: Fetch booking details
  const { default: postgres } = await import('postgres');
  const sql = postgres(DB_URL);
  
  const bookingRows = await sql<{
    booking_id: string;
    provider_id: string;
    client_id: string;
    start_time: string;
    end_time: string;
    status: string;
    gcal_sync_status: string;
    gcal_provider_event_id: string | null;
  }[]>`
    SELECT booking_id, provider_id, client_id, start_time, end_time, status, gcal_sync_status, gcal_provider_event_id
    FROM bookings
    WHERE booking_id = ${BOOKING_ID}::uuid
  `;
  
  const booking = bookingRows[0];
  if (!booking) {
    console.log('❌ Booking not found');
    await sql.end();
    return;
  }
  
  console.log('✅ Booking found:');
  console.log(`   Start: ${new Date(booking.start_time).toISOString()}`);
  console.log(`   End: ${new Date(booking.end_time).toISOString()}`);
  console.log(`   Status: ${booking.status}`);
  console.log(`   GCal Status: ${booking.gcal_sync_status}`);
  
  // Step 2: Check GCal credentials
  if (!GCAL_ACCESS_TOKEN || GCAL_ACCESS_TOKEN === 'ya29.xxx') {
    console.log('\n⚠️  GCAL_ACCESS_TOKEN not configured or is placeholder');
    console.log('   GCal sync cannot proceed without valid credentials');
    console.log('\n📋 To configure GCal:');
    console.log('   1. Go to Google Cloud Console');
    console.log('   2. Enable Calendar API');
    console.log('   3. Create OAuth 2.0 credentials');
    console.log('   4. Get access token and refresh token');
    console.log('   5. Add to .env file');
    await sql.end();
    return;
  }
  
  if (!GCAL_PROVIDER_CALENDAR_ID) {
    console.log('\n⚠️  GCAL_PROVIDER_CALENDAR_ID not configured');
    await sql.end();
    return;
  }
  
  console.log('\n✅ GCal credentials found');
  console.log(`   Calendar: ${GCAL_PROVIDER_CALENDAR_ID}`);
  
  // Step 3: Create GCal event
  const eventPayload = {
    summary: `Cita Médica - Test Cliente Nuevo`,
    description: `Consulta General\nBooking ID: ${booking.booking_id}`,
    start: {
      dateTime: new Date(booking.start_time).toISOString(),
      timeZone: 'America/Mexico_City',
    },
    end: {
      dateTime: new Date(booking.end_time).toISOString(),
      timeZone: 'America/Mexico_City',
    },
    status: 'confirmed',
  };
  
  console.log('\n📤 Creating GCal event...');
  
  try {
    const url = `https://www.googleapis.com/calendar/v3/calendars/${encodeURIComponent(GCAL_PROVIDER_CALENDAR_ID)}/events`;
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${GCAL_ACCESS_TOKEN}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(eventPayload),
    });
    
    if (!response.ok) {
      const errorText = await response.text();
      console.log(`❌ GCal API error: ${response.status}`);
      console.log(errorText);
      
      // Update booking status to 'failed'
      await sql`
        UPDATE bookings
        SET gcal_sync_status = 'failed'
        WHERE booking_id = ${BOOKING_ID}::uuid
      `;
      console.log('   Updated booking gcal_sync_status to "failed"');
      await sql.end();
      return;
    }
    
    const eventData = await response.json();
    console.log('✅ GCal event created successfully!');
    console.log(`   Event ID: ${eventData.id}`);
    console.log(`   HTML Link: ${eventData.htmlLink}`);
    console.log(`   Status: ${eventData.status}`);
    
    // Update booking with GCal event ID
    await sql`
      UPDATE bookings
      SET gcal_sync_status = 'synced',
          gcal_provider_event_id = ${eventData.id ?? ''},
          gcal_last_sync = NOW()
      WHERE booking_id = ${BOOKING_ID}::uuid
    `;
    console.log('   Updated booking gcal_sync_status to "synced"');
    
  } catch (e) {
    console.error(`❌ GCal sync failed: ${e instanceof Error ? e.message : String(e)}`);
    await sql`
      UPDATE bookings
      SET gcal_sync_status = 'failed'
      WHERE booking_id = ${BOOKING_ID}::uuid
    `;
  }
  
  await sql.end();
}

main().catch(e => {
  console.error('Fatal:', e);
  process.exit(1);
});
